File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        
        new_head = head.next
        cur = head
        prev = None
        while cur is not None and cur.next is not None:
            p1, p2 = cur, cur.next
            cur = cur.next.next
            p2.next = p1
            p1.next = cur
            if prev:
                prev.next = p2
            prev = p1    
            
        return new_head   

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    next := head.Next.Next
    p1, p2 := head, head.Next
    p2.Next = p1
    p1.Next = swapPairs(next)
    return p2
}